File: sdk_v2/cs/src/Configuration.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Microsoft.AI.Foundry.Local;

public class Configuration
{
    /// <summary>
    /// Your application name. MUST be set to a valid name.
    /// </summary>
    public required string AppName { get; set; }

    /// <summary>
    /// Application data directory.
    /// Default: {home}/.{appname}, where {home} is the user's home directory and {appname} is the AppName value.
    /// </summary>
    public string? AppDataDir { get; init; }

    /// <summary>
    /// Model cache directory.
    /// Default: {appdata}/cache/models, where {appdata} is the AppDataDir value.
    /// </summary>
    public string? ModelCacheDir { get; init; }

    /// <summary>
    /// Log directory.
    /// Default: {appdata}/logs
    /// </summary>
    public string? LogsDir { get; init; }

    /// <summary>
    /// Logging level.
    /// Valid values are: Verbose, Debug, Information, Warning, Error, Fatal.
    /// Default: LogLevel.Warning
    /// </summary>
    public LogLevel LogLevel { get; init; } = LogLevel.Warning;

    /// <summary>
    /// Enable manual execution provider download mode. Only meaningful if using WinML.
    /// 
    /// Default: false
    /// 
    /// When false, EPs are downloaded automatically in the background when FoundryLocalManager is created.
    /// When true, EPs are downloaded when FoundryLocalManager.EnsureEpsDownloadedAsync or GetCatalogAsync are called.
    ///
    /// Once an EP is downloaded it will not be re-downloaded unless a new version is available.
    /// </summary>
    // DISABLED: We want to make sure this is required before making it public as supporting this complicates the
    //           Core implementation. Can be specified via AdditionalSettings if needed for testing.
    // public bool ManualEpDownload { get; init; } 

    /// <summary>
    /// Optional configuration for the built-in web service.
    /// NOTE: This is not included in all builds.
    /// </summary>
    public WebService? Web { get; init; }

    /// <summary>
    /// Additional settings that Foundry Local Core can consume.
    /// Keys and values are strings.
    /// </summary>
    public IDictionary<string, string>? AdditionalSettings { get; init; }

    /// <summary>
    /// Configuration settings if the optional web service is used.
    /// </summary>
    public class WebService
    {
        /// <summary>
        /// Url/s to bind to the web service when <see cref="FoundryLocalManager.StartWebServiceAsync"/> is called.
        /// After startup, <see cref="FoundryLocalManager.Urls"/> will contain the actual URL/s the service is listening on.
        /// 
        /// Default: 127.0.0.1:0, which binds to a random ephemeral port.
        /// Multiple URLs can be specified as a semi-colon separated list.
        /// </summary>
        public string? Urls { get; init; }

        /// <summary>
        /// If the web service is running in a separate process, it will be accessed using this URI.
        /// </summary>
        /// <remarks>
        /// Both processes should be using the same version of the SDK. If a random port is assigned when creating
        /// the web service in the external process the actual port must be provided here.
        /// </remarks> 
        public Uri? ExternalUrl { get; init; }
    }

    internal void Validate()
    {
        if (string.IsNullOrEmpty(AppName))
        {
            throw new ArgumentException("Configuration AppName must be set to a valid application name.");
        }

        if (AppName.IndexOfAny(Path.GetInvalidFileNameChars()) >= 0)
        {
            throw new ArgumentException("Configuration AppName value contains invalid characters.");
        }


        if (Web?.ExternalUrl?.Port == 0)
        {
            throw new ArgumentException("Configuration Web.ExternalUrl has invalid port of 0.");
        }
    }

    internal Dictionary<string, string> AsDictionary()
    {
        if (string.IsNullOrEmpty(AppName))
        {
            throw new FoundryLocalException(
                "Configuration AppName must be set to a valid application name.");
        }

        var configValues = new Dictionary<string, string>
        {
            { "AppName", AppName },
            { "LogLevel", LogLevel.ToString() }
        };

        if (!string.IsNullOrEmpty(AppDataDir))
        {
            configValues.Add("AppDataDir", AppDataDir);
        }

        if (!string.IsNullOrEmpty(ModelCacheDir))
        {
            configValues.Add("ModelCacheDir", ModelCacheDir);
        }

        if (!string.IsNullOrEmpty(LogsDir))
        {
            configValues.Add("LogsDir", LogsDir);
        }

        //configValues.Add("ManualEpDownload", ManualEpDownload.ToString());

        if (Web != null)
        {
            if (Web.Urls != null)
            {
                configValues["WebServiceUrls"] = Web.Urls;
            }
        }

        // Emit any additional settings.
        if (AdditionalSettings != null)
        {
            foreach (var kvp in AdditionalSettings)
            {
                if (string.IsNullOrEmpty(kvp.Key))
                {
                    continue; // skip empty keys
                }
                configValues[kvp.Key] = kvp.Value ?? string.Empty;
            }
        }

        return configValues;
    }
}

File: sdk_v2/rust/src/configuration.rs
use std::collections::HashMap;

use crate::error::{FoundryLocalError, Result};

/// Log level for the Foundry Local service.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LogLevel {
    Trace,
    Debug,
    Info,
    Warn,
    Error,
    Fatal,
}

impl LogLevel {
    /// Returns the string value expected by the native core library.
    fn as_core_str(&self) -> &'static str {
        match self {
            Self::Trace => "Verbose",
            Self::Debug => "Debug",
            Self::Info => "Information",
            Self::Warn => "Warning",
            Self::Error => "Error",
            Self::Fatal => "Fatal",
        }
    }
}

/// User-facing configuration for initializing the Foundry Local SDK.
#[derive(Debug, Clone, Default)]
#[non_exhaustive]
pub struct FoundryLocalConfig {
    pub app_name: String,
    pub app_data_dir: Option<String>,
    pub model_cache_dir: Option<String>,
    pub logs_dir: Option<String>,
    pub log_level: Option<LogLevel>,
    pub web_service_urls: Option<String>,
    pub service_endpoint: Option<String>,
    pub library_path: Option<String>,
    pub additional_settings: Option<HashMap<String, String>>,
}

impl FoundryLocalConfig {
    /// Create a new configuration with the given application name.
    ///
    /// All other fields default to `None`. Use the struct update syntax to
    /// override specific options:
    ///
    /// ```ignore
    /// let config = FoundryLocalConfig {
    ///     log_level: Some(LogLevel::Debug),
    ///     ..FoundryLocalConfig::new("my_app")
    /// };
    /// ```
    pub fn new(app_name: impl Into<String>) -> Self {
        Self {
            app_name: app_name.into(),
            ..Self::default()
        }
    }
}

/// Internal configuration object that converts [`FoundryLocalConfig`] into the
/// parameter map expected by the native core library.
#[derive(Debug, Clone)]
pub(crate) struct Configuration {
    pub params: HashMap<String, String>,
}

impl Configuration {
    /// Build a [`Configuration`] from the user-facing [`FoundryLocalConfig`].
    ///
    /// # Errors
    ///
    /// Returns [`FoundryLocalError::InvalidConfiguration`] when `app_name` is
    /// empty or blank.
    pub fn new(config: FoundryLocalConfig) -> Result<Self> {
        let app_name = config.app_name.trim().to_string();
        if app_name.is_empty() {
            return Err(FoundryLocalError::InvalidConfiguration {
                reason: "app_name must be set and non-empty".into(),
            });
        }

        let mut params = HashMap::new();
        params.insert("AppName".into(), app_name);

        if let Some(v) = config.app_data_dir {
            params.insert("AppDataDir".into(), v);
        }
        if let Some(v) = config.model_cache_dir {
            params.insert("ModelCacheDir".into(), v);
        }
        if let Some(v) = config.logs_dir {
            params.insert("LogsDir".into(), v);
        }
        if let Some(level) = config.log_level {
            params.insert("LogLevel".into(), level.as_core_str().into());
        }
        if let Some(v) = config.web_service_urls {
            params.insert("WebServiceUrls".into(), v);
        }
        if let Some(v) = config.service_endpoint {
            params.insert("WebServiceExternalUrl".into(), v);
        }
        if let Some(v) = config.library_path {
            params.insert("FoundryLocalCorePath".into(), v);
        }
        if let Some(extra) = config.additional_settings {
            for (k, v) in extra {
                params.insert(k, v);
            }
        }

        Ok(Self { params })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_config() {
        let cfg = FoundryLocalConfig {
            log_level: Some(LogLevel::Debug),
            ..FoundryLocalConfig::new("TestApp")
        };
        let c = Configuration::new(cfg).unwrap();
        assert_eq!(c.params["AppName"], "TestApp");
        assert_eq!(c.params["LogLevel"], "Debug");
    }

    #[test]
    fn empty_app_name_fails() {
        let cfg = FoundryLocalConfig {
            app_name: "  ".into(),
            ..FoundryLocalConfig::default()
        };
        assert!(Configuration::new(cfg).is_err());
    }
}
